### PR TITLE
[SPARK-15657][SQL] RowEncoder should validate the data type of input object

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

This PR improves the error handling of `RowEncoder`. When we create a `RowEncoder` with a given schema, we should validate the data type of input object. e.g. we should throw an exception when a field is boolean but is declared as a string column.

This PR also removes the support to use `Product` as a valid external type of struct type.  This support is added at https://github.com/apache/spark/pull/9712, but is incomplete, e.g. nested product, product in array are both not working.  However, we never officially support this feature and I think it's ok to ban it.
## How was this patch tested?

new tests in `RowEncoderSuite`.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @marmbrus  @yhuai @viirya

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 1]
```diff
@@ -721,8 +716,55 @@ case class GetExternalRowField(
...
+          .map(cls => s"$obj instanceof ${cls.getName}").mkString(" || ")
+      case _: ArrayType =>
+        s"$obj instanceof ${classOf[Seq[_]].getName} || $obj.getClass().isArray()"
+      case _ =>
+        s"$obj instanceof ${ctx.boxedType(dataType)}"
+    }
+
+    val code = s"""
+      ${input.code}
+      ${ctx.javaType(dataType)} ${ev.value} = ${ctx.defaultValue(dataType)};
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: `ctx.boxedType` here too?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala [hunk 2]
```diff
@@ -721,8 +716,55 @@ case class GetExternalRowField(
...
      """
     ev.copy(code = code, isNull = "false")
   }
 }
+
+/**
+ * Validates the actual data type of input expression at runtime.  If it doesn't match the
+ * expectation, throw an exception.
+ */
+case class ValidateExternalType(child: Expression, expected: DataType)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: Previously we completely rely on the correctness of provided schema. If the data doesn't match provided schema, we should see value converting error. I think this `ValidateExternalType` can improve the experience of error handling. However, it looks like a performance regression as we do extra checking here. If we can trust the provided schema, should we do this checking?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala (hunk 1 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: This is by means, we can't cast an object to int directly, but have to cast to boxed int first.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala (hunk 2 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: The problem is we can't trust it.... When users call `createDataFrame(rows, schema)`, we should definitely validate the passed-in rows. I think performance doesn't matter too much here, as this only happens at the beginning of the data flow. One potential issue may be that, `Dataset.map` can return row and users will provide a schema we should trust. However, I don't think we should expose `RowEncoder` to users and `Dataset.map` should never return a row.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: hmm, for the trust, it should be said that we leave the responsibility of data correctness to users. Actually when the data is in wrong type, we will not get wrong result. There is of course an exception regarding data converting will be thrown. However, better error handling is always good if as you said the performance is not big issue here.

**Pull request closed without merging**